rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,185 @@
+
+* The actual binary distribution zipfile
+* A checksum file containing the SHA256 of the zipfile (formatted as per
+ `sha256sum` tool)
+* A signature file containing a binary, detached GPG signature for
+  checksum file.
+
+The Teleport provider release process currently produces a compressed tarball
+by default, and this should still be considered the primary artefact produced
+by the build, at least as long as Houston is expected to serve downloads of
+the Terraform provider.
+
+By analogy with the main Teleport release system, the registry-compatible
+provider package will be created during release promotion, with a build task
+that will:
+
+1. For each release tarball (one each for multiple release platforms):
+    1. repack the tarball into an appropriately formatted & named zipfile
+    2. generate the checksum file from the zipfile
+    3. generate the signature file by signing the checksum file
+2. Upload all generated files to the AWS bucket that backs the registry
+3. Update the registry index to include the newly released providers (see
+   _Populating the registry_, below)
+
+I expect this to be implemented as a small Go program, rather than a shell
+script. Using small Go programs for build tasks has shown itself to be a
+useful, flexible and above all _legible_ tool for scripting build tasks in
+the mainline Teleport repository, and we should continue that pattern in
+`teleport-plugins`.
+
+#### Build Security issues
+
+This process will require that the `teleport-plugins` promotion process will
+need new secrets:
+
+1. Read access to a signing key, to sign the package (see note below)
+2. Write access to the AWS bucket storing the signed packages for distribution
+
+_**Note:**_ We currently expect to use the RPM signing key as the Terraform
+signing key as it already exists and is already used for a similar purpose.
+That said, there is no technical reason why it _must_ be the RPM signing
+key. As long as the private key used to sign the package agrees with the a
+public key exposed via the registry, the signature should be deemed valid
+by Terraform.
+
+### Populating the registry
+
+Populating the registry is a two-part problem:
+
+First, we must create a new, back-filled registry containing all of the
+currently available Teleport terraform providers.
+
+This appears to be pretty straightforward, as I already have a proof-of-
+concept program that scans the release plugins AWS bucket and repacks all
+of the available releases, generating a TypeScript dictionary that can be
+embedded directly in the front-end registry code.
+
+Secondly, we need a mechanism to _maintain_ the registry as new releases
+are produced.
+
+_This_ is the major open question in this RFD - _how will our promotion
+process communicate to our custom registry that a new release exists?_
+
+#### Options
+
+1. Fully manual process
+2. Promotion process generates an index file that can be used by the
+   registry in `gravitational/next`, either by
```

Comment:
I like this solution as a "good enough" solution.

Upsides:
* No need to host extra logic in next/somewhere else.
* Consistent with @fheinecke's plans in https://github.com/gravitational/teleport/pull/10746 and our other repos

Downsides:

* I haven't looked at the TF index format yet.  If it explicitly has a "latest" or other floating tags, we'll end up in a https://github.com/gravitational/cloud/issues/958 situation
* Concurrency concerns -- we don't want two builds to be rebuilding/uploading an index at the same time.  We need some sort of lease/mutex to do this properly.  That said, we've gotten pretty far with only drone's concurrency controls.